src/main/java/org/opentripplanner/graph_builder/linking/ToStreetEdgeLinker.java

```diff
@@ -0,0 +1,91 @@
+package org.opentripplanner.graph_builder.linking;
+
+import org.locationtech.jts.geom.LineString;
+import org.locationtech.jts.linearref.LinearLocation;
+import org.opentripplanner.routing.core.RoutingRequest;
+import org.opentripplanner.routing.core.TraverseMode;
+import org.opentripplanner.routing.edgetype.StreetEdge;
+import org.opentripplanner.routing.graph.Vertex;
+import org.opentripplanner.routing.location.TemporaryStreetLocation;
+import org.opentripplanner.routing.vertextype.StreetVertex;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Tries to link vertices to all closest edges in graph. If possible, it maset links directly to vertexes
+ * via {@link EdgesMaker}, or else it delegates creating links to {@link ToEdgeLinker}
+ */
+public class ToStreetEdgeLinker {
+
+    private final ToEdgeLinker toEdgeLinker;
+
+    private final EdgesToLinkFinder edgesToLinkFinder;
+
+    private final LinkingGeoTools linkingGeoTools;
+
+    private final EdgesMaker edgesMaker;
+
+    public ToStreetEdgeLinker(ToEdgeLinker toEdgeLinker, EdgesToLinkFinder edgesToLinkFinder, LinkingGeoTools linkingGeoTools,
+                              EdgesMaker edgesMaker) {
+        this.toEdgeLinker = toEdgeLinker;
+        this.edgesToLinkFinder = edgesToLinkFinder;
+        this.linkingGeoTools = linkingGeoTools;
+        this.edgesMaker = edgesMaker;
+    }
+
+    /**
+     * Temporarily link this vertex into the graph
+     */
+    public boolean linkTemporarily(TemporaryStreetLocation vertex, TraverseMode traverseMode, RoutingRequest options) {
+        List<StreetEdge> streetEdges = edgesToLinkFinder.findEdgesToLink(vertex, traverseMode);
+        streetEdges.forEach(edge -> linkTemporarilyToEdge(vertex, edge, options));
+        return !streetEdges.isEmpty();
+    }
+
+    /**
+     * Permanently link this vertex into the graph
+     */
+    public boolean linkPermanently(Vertex vertex, TraverseMode traverseMode) {
+        List<StreetEdge> streetEdges = edgesToLinkFinder.findEdgesToLink(vertex, traverseMode);
+        streetEdges.forEach(edge -> linkPermanentlyToEdge(vertex, edge));
+        return !streetEdges.isEmpty();
+    }
+
+    private void linkTemporarilyToEdge(TemporaryStreetLocation vertex, StreetEdge edge, RoutingRequest options) {
+        LineString orig = edge.getGeometry();
+        LinearLocation ll = linkingGeoTools.findLocationClosestToVertex(vertex, orig);
+        Optional<Vertex> maybeVertexToLinkTo = maybeFindVertexToLinkTo(edge, orig, ll);
+        if (maybeVertexToLinkTo.isPresent()) {
+            edgesMaker.makeTemporaryEdges(vertex, maybeVertexToLinkTo.get());
+        } else {
+            //This throws runtime TrivialPathException if same edge is split in origin and destination link
```

Comment:
Wiem, że to runtime, ale może dodaj `throws TrivialPathException` do sygnatury aż do handlera? Ja jestem za tym, żeby nie trakować runtime exception inaczej bo one i tak są złem wcielonym

Reply:
poprawione